File: packages/python/goldenmatch/goldenmatch/core/config_edits.py
"""Shared ConfigEdit lever vocabulary.

A small, closed edit language: each edit is a validated mutation of a
``GoldenMatchConfig`` with a human label. ``apply()`` returns ``None`` for a
no-op or an edit that would produce an invalid config (the caller skips it).

This vocabulary is the single lever language shared by:

- the agentic optimizer's proposers (``core/config_optimizer.py``) — each edit
  becomes one scored candidate; and
- the controller's ``LLMRefitPolicy`` (``core/autoconfig_policy.py``) — the LLM
  repair folds a list of edits onto the current config.

Living in its own module keeps the dependency graph acyclic (the optimizer
imports the policy, so the vocabulary can't live in either).
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from goldenmatch.config.schemas import BlockingKeyConfig, GoldenMatchConfig

__all__ = [
    "ConfigEdit",
    "ThresholdShift",
    "ScorerSwap",
    "BlockingStrategyEdit",
    "WeightShift",
    "MatchkeyTypeSwap",
    "BlockingKeyEdit",
    "edit_from_spec",
    "parse_llm_edits",
    "fold_edits",
]

_PERTURBABLE_TYPES = ("weighted", "probabilistic")


def _clamp(x: float, lo: float = 0.0, hi: float = 1.0) -> float:
    return max(lo, min(hi, x))


def _perturbable_matchkeys(config: GoldenMatchConfig) -> list:
    return [
        mk for mk in config.get_matchkeys()
        if getattr(mk, "type", None) in _PERTURBABLE_TYPES and mk.threshold is not None
    ]


class ConfigEdit(Protocol):
    @property
    def label(self) -> str: ...

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None: ...


def _revalidate(cfg: GoldenMatchConfig) -> GoldenMatchConfig | None:
    try:
        return GoldenMatchConfig.model_validate(cfg.model_dump())
    except Exception:  # noqa: BLE001 - invalid edit -> skip
        return None


@dataclass(frozen=True)
class ThresholdShift:
    """Shift every perturbable matchkey threshold by ``delta`` (clamped)."""

    delta: float

    @property
    def label(self) -> str:
        return "baseline" if self.delta == 0.0 else f"threshold{self.delta:+.2f}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        if not _perturbable_matchkeys(config):
            return config if self.delta == 0.0 else None
        cfg = config.model_copy(deep=True)
        changed = False
        for mk in cfg.get_matchkeys():
            if getattr(mk, "type", None) in _PERTURBABLE_TYPES and mk.threshold is not None:
                new_t = _clamp(mk.threshold + self.delta)
                if new_t != mk.threshold:
                    mk.threshold = new_t
                    changed = True
        if self.delta == 0.0:
            return cfg  # baseline: valid, no change
        return cfg if changed else None


@dataclass(frozen=True)
class ScorerSwap:
    """Swap one matchkey field's scorer."""

    matchkey: str
    field: str
    scorer: str

    @property
    def label(self) -> str:
        return f"scorer:{self.field}={self.scorer}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        cfg = config.model_copy(deep=True)
        changed = False
        for mk in cfg.get_matchkeys():
            if mk.name != self.matchkey:
                continue
            for f in (mk.fields or []):
                if f.field == self.field and f.scorer != self.scorer:
                    f.scorer = self.scorer
                    changed = True
        if not changed:
            return None
        return _revalidate(cfg)


@dataclass(frozen=True)
class BlockingStrategyEdit:
    """Change the blocking strategy (keys preserved)."""

    strategy: str

    @property
    def label(self) -> str:
        return f"blocking:{self.strategy}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        if config.blocking is None or config.blocking.strategy == self.strategy:
            return None
        cfg = config.model_copy(update={
            "blocking": config.blocking.model_copy(update={"strategy": self.strategy}),
        })
        return _revalidate(cfg)


@dataclass(frozen=True)
class WeightShift:
    """Reweight one field of a weighted matchkey (floor 0.0)."""

    matchkey: str
    field: str
    delta: float

    @property
    def label(self) -> str:
        return f"weight:{self.field}{self.delta:+.2f}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        cfg = config.model_copy(deep=True)
        changed = False
        for mk in cfg.get_matchkeys():
            if mk.name != self.matchkey or getattr(mk, "type", None) != "weighted":
                continue
            for f in (mk.fields or []):
                if f.field == self.field and f.weight is not None:
                    new_w = max(0.0, f.weight + self.delta)
                    if new_w != f.weight:
                        f.weight = new_w
                        changed = True
        if not changed:
            return None
        return _revalidate(cfg)


@dataclass(frozen=True)
class MatchkeyTypeSwap:
    """Swap a matchkey between ``weighted`` and ``probabilistic``.

    A weighted matchkey already carries everything probabilistic needs (each
    field has a scorer); going the other way we backfill a threshold and uniform
    per-field weights so the weighted invariant holds.
    """

    matchkey: str
    target_type: str  # "weighted" | "probabilistic"

    @property
    def label(self) -> str:
        return f"mktype:{self.matchkey}={self.target_type}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        if self.target_type not in _PERTURBABLE_TYPES:
            return None
        cfg = config.model_copy(deep=True)
        changed = False
        for mk in cfg.get_matchkeys():
            cur = getattr(mk, "type", None)
            if mk.name != self.matchkey or cur == self.target_type or cur not in _PERTURBABLE_TYPES:
                continue
            mk.type = self.target_type
            mk.comparison = None
            if self.target_type == "weighted":
                if mk.threshold is None:
                    mk.threshold = mk.link_threshold if mk.link_threshold is not None else 0.5
                for f in (mk.fields or []):
                    if f.weight is None:
                        f.weight = 1.0
            changed = True
        if not changed:
            return None
        return _revalidate(cfg)


@dataclass(frozen=True)
class BlockingKeyEdit:
    """Add or remove a blocking key, identified by its field set + transforms."""

    action: str  # "add" | "remove"
    fields: tuple[str, ...]
    transforms: tuple[str, ...] = ()

    @property
    def label(self) -> str:
        return f"block-{self.action}:{'+'.join(self.fields)}"

    def apply(self, config: GoldenMatchConfig) -> GoldenMatchConfig | None:
        if config.blocking is None or self.action not in ("add", "remove") or not self.fields:
            return None
        target, ttx = list(self.fields), list(self.transforms)
        keys = list(config.blocking.keys or [])
        exists = any(k.fields == target and (k.transforms or []) == ttx for k in keys)
        if self.action == "add":
            if exists:
                return None
            keys = [*keys, BlockingKeyConfig(fields=target, transforms=ttx)]
        else:
            if not exists:
                return None
            keys = [k for k in keys if not (k.fields == target and (k.transforms or []) == ttx)]
        cfg = config.model_copy(update={"blocking": config.blocking.model_copy(update={"keys": keys})})
        return _revalidate(cfg)


def edit_from_spec(spec: object) -> ConfigEdit | None:
    """Map one LLM-emitted edit spec (a JSON object) to a ``ConfigEdit``.

    Returns ``None`` for an unknown op or a malformed spec — the closed
    vocabulary is the only thing the LLM can drive, so junk is dropped, not run.
    """
    if not isinstance(spec, dict):
        return None
    op = spec.get("op")
    try:
        if op == "threshold_shift":
            return ThresholdShift(float(spec["delta"]))
        if op == "scorer_swap":
            return ScorerSwap(str(spec["matchkey"]), str(spec["field"]), str(spec["scorer"]))
        if op == "blocking_strategy":
            return BlockingStrategyEdit(str(spec["strategy"]))
        if op == "weight_shift":
            return WeightShift(str(spec["matchkey"]), str(spec["field"]), float(spec["delta"]))
        if op == "matchkey_type":
            return MatchkeyTypeSwap(str(spec["matchkey"]), str(spec["target_type"]))
        if op == "blocking_key":
            return BlockingKeyEdit(
                str(spec["action"]),
                tuple(spec["fields"]),
                tuple(spec.get("transforms", ())),
            )
    except (KeyError, TypeError, ValueError):
        return None
    return None


def parse_llm_edits(payload: object) -> list[ConfigEdit]:
    """Parse an LLM response into a list of validated ``ConfigEdit``s.

    ``{"action": "stop"}`` / ``{"action": "satisfied"}`` (or any non-list
    ``edits``) yields an empty list.
    """
    if not isinstance(payload, dict) or payload.get("action") in ("stop", "satisfied"):
        return []
    raw = payload.get("edits")
    if not isinstance(raw, list):
        return []
    return [e for e in (edit_from_spec(s) for s in raw) if e is not None]


def fold_edits(config: GoldenMatchConfig, edits: list[ConfigEdit]) -> GoldenMatchConfig:
    """Apply ``edits`` in sequence onto ``config``, skipping any that don't
    apply (return ``None``). Returns the folded config — equal to the input
    when every edit was a no-op. Used by the controller's single-trajectory LLM
    repair, where a list of edits composes into one new config."""
    cur = config
    for edit in edits:
        nxt = edit.apply(cur)
        if nxt is not None:
            cur = nxt
    return cur

File: packages/typescript/goldenmatch/src/core/config-edits.ts
/**
 * config-edits.ts -- shared ConfigEdit lever vocabulary.
 *
 * Port of Python `goldenmatch/core/config_edits.py`: a small, closed edit
 * language where each edit is a validated transform of a `GoldenMatchConfig`
 * with a human label. `apply()` returns `null` for a no-op or an edit that
 * would produce an invalid config (the caller skips it).
 *
 * TS configs are immutable, so edits build new objects instead of pydantic's
 * deep-copy-then-mutate; pydantic revalidation maps to explicit checks
 * (`VALID_SCORERS`, the blocking-strategy union). Parity is fixture-verified:
 * tests/parity/config-edits.test.ts replays Python-emitted edit specs and
 * compares semantic projections of the results.
 */
import type {
  BlockingConfig,
  BlockingKeyConfig,
  GoldenMatchConfig,
  MatchkeyConfig,
} from "./types.js";
import { getMatchkeys, VALID_SCORERS } from "./types.js";

const PERTURBABLE_TYPES = new Set<string>(["weighted", "probabilistic"]);

const VALID_BLOCKING_STRATEGIES = new Set<string>([
  "static",
  "adaptive",
  "sorted_neighborhood",
  "multi_pass",
  "ann",
  "canopy",
  "ann_pairs",
  "learned",
]);

export interface ConfigEdit {
  readonly label: string;
  apply(config: GoldenMatchConfig): GoldenMatchConfig | null;
}

function clamp(x: number, lo = 0.0, hi = 1.0): number {
  return Math.max(lo, Math.min(hi, x));
}

function thresholdOf(mk: MatchkeyConfig): number | undefined {
  return (mk as { threshold?: number }).threshold;
}

/** Write the matchkeys back to whichever property held them. */
function withMatchkeys(
  config: GoldenMatchConfig,
  mks: readonly MatchkeyConfig[],
): GoldenMatchConfig {
  if (config.matchkeys !== undefined) return { ...config, matchkeys: mks };
  return { ...config, matchSettings: mks };
}

function arrayEq(a: readonly string[], b: readonly string[]): boolean {
  return a.length === b.length && a.every((v, i) => v === b[i]);
}

/**
 * Pydantic-revalidation parity for blocking edits (schemas.py BlockingConfig
 * validator): static/adaptive require `keys` (or `sub_block_keys`);
 * multi_pass requires `keys` or `passes`. An edit that would leave the
 * blocking config invalid is skipped (returns null), same as Python.
 */
function blockingValid(b: BlockingConfig): boolean {
  const keys = b.keys ?? [];
  const subBlockKeys =
    (b as { subBlockKeys?: readonly unknown[] }).subBlockKeys ?? [];
  const passes = (b as { passes?: readonly unknown[] }).passes ?? [];
  if (
    (b.strategy === "static" || b.strategy === "adaptive") &&
    keys.length === 0 &&
    subBlockKeys.length === 0
  ) {
    return false;
  }
  if (b.strategy === "multi_pass" && keys.length === 0 && passes.length === 0) {
    return false;
  }
  return true;
}

/** Shift every perturbable matchkey threshold by `delta` (clamped). */
export class ThresholdShift implements ConfigEdit {
  constructor(readonly delta: number) {}

  get label(): string {
    if (this.delta === 0.0) return "baseline";
    const sign = this.delta >= 0 ? "+" : "";
    return `threshold${sign}${this.delta.toFixed(2)}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    const mks = getMatchkeys(config);
    const perturbable = mks.filter(
      (mk) => PERTURBABLE_TYPES.has(mk.type) && thresholdOf(mk) !== undefined,
    );
    if (perturbable.length === 0) {
      return this.delta === 0.0 ? config : null;
    }
    let changed = false;
    const next = mks.map((mk) => {
      const t = thresholdOf(mk);
      if (PERTURBABLE_TYPES.has(mk.type) && t !== undefined) {
        const nt = clamp(t + this.delta);
        if (nt !== t) {
          changed = true;
          return { ...mk, threshold: nt } as MatchkeyConfig;
        }
      }
      return mk;
    });
    if (this.delta === 0.0) return withMatchkeys(config, next);
    return changed ? withMatchkeys(config, next) : null;
  }
}

/** Swap one matchkey field's scorer. */
export class ScorerSwap implements ConfigEdit {
  constructor(
    readonly matchkey: string,
    readonly field: string,
    readonly scorer: string,
  ) {}

  get label(): string {
    return `scorer:${this.field}=${this.scorer}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    // Pydantic-revalidation parity: an unknown scorer is an invalid edit.
    if (!(VALID_SCORERS as ReadonlySet<string>).has(this.scorer)) return null;
    let changed = false;
    const next = getMatchkeys(config).map((mk) => {
      if (mk.name !== this.matchkey) return mk;
      let mkChanged = false;
      const fields = mk.fields.map((f) => {
        if (f.field === this.field && f.scorer !== this.scorer) {
          mkChanged = true;
          return { ...f, scorer: this.scorer };
        }
        return f;
      });
      if (!mkChanged) return mk;
      changed = true;
      return { ...mk, fields } as MatchkeyConfig;
    });
    return changed ? withMatchkeys(config, next) : null;
  }
}

/** Change the blocking strategy (keys preserved). */
export class BlockingStrategyEdit implements ConfigEdit {
  constructor(readonly strategy: string) {}

  get label(): string {
    return `blocking:${this.strategy}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    if (
      config.blocking === undefined ||
      config.blocking.strategy === this.strategy ||
      !VALID_BLOCKING_STRATEGIES.has(this.strategy)
    ) {
      return null;
    }
    const blocking = {
      ...config.blocking,
      strategy: this.strategy as BlockingConfig["strategy"],
    };
    if (!blockingValid(blocking)) return null;
    return { ...config, blocking };
  }
}

/** Reweight one field of a weighted matchkey (floor 0.0). */
export class WeightShift implements ConfigEdit {
  constructor(
    readonly matchkey: string,
    readonly field: string,
    readonly delta: number,
  ) {}

  get label(): string {
    const sign = this.delta >= 0 ? "+" : "";
    return `weight:${this.field}${sign}${this.delta.toFixed(2)}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    let changed = false;
    const next = getMatchkeys(config).map((mk) => {
      if (mk.name !== this.matchkey || mk.type !== "weighted") return mk;
      let mkChanged = false;
      const fields = mk.fields.map((f) => {
        if (f.field === this.field && f.weight !== undefined) {
          const nw = Math.max(0.0, f.weight + this.delta);
          if (nw !== f.weight) {
            mkChanged = true;
            return { ...f, weight: nw };
          }
        }
        return f;
      });
      if (!mkChanged) return mk;
      changed = true;
      return { ...mk, fields } as MatchkeyConfig;
    });
    return changed ? withMatchkeys(config, next) : null;
  }
}

/**
 * Swap a matchkey between `weighted` and `probabilistic`. Going to weighted we
 * backfill a threshold (`linkThreshold` else 0.5) and uniform per-field
 * weights so the weighted invariant holds.
 */
export class MatchkeyTypeSwap implements ConfigEdit {
  constructor(
    readonly matchkey: string,
    readonly targetType: string,
  ) {}

  get label(): string {
    return `mktype:${this.matchkey}=${this.targetType}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    if (!PERTURBABLE_TYPES.has(this.targetType)) return null;
    let changed = false;
    const next = getMatchkeys(config).map((mk) => {
      if (
        mk.name !== this.matchkey ||
        mk.type === this.targetType ||
        !PERTURBABLE_TYPES.has(mk.type)
      ) {
        return mk;
      }
      changed = true;
      if (this.targetType === "weighted") {
        const link = (mk as { linkThreshold?: number }).linkThreshold;
        const threshold = thresholdOf(mk) ?? link ?? 0.5;
        const fields = mk.fields.map((f) =>
          f.weight === undefined ? { ...f, weight: 1.0 } : f,
        );
        return { ...mk, type: "weighted", threshold, fields } as MatchkeyConfig;
      }
      return { ...mk, type: "probabilistic" } as MatchkeyConfig;
    });
    return changed ? withMatchkeys(config, next) : null;
  }
}

/** Add or remove a blocking key, identified by its field set + transforms. */
export class BlockingKeyEdit implements ConfigEdit {
  constructor(
    readonly action: string,
    readonly fields: readonly string[],
    readonly transforms: readonly string[] = [],
  ) {}

  get label(): string {
    return `block-${this.action}:${this.fields.join("+")}`;
  }

  apply(config: GoldenMatchConfig): GoldenMatchConfig | null {
    if (
      config.blocking === undefined ||
      (this.action !== "add" && this.action !== "remove") ||
      this.fields.length === 0
    ) {
      return null;
    }
    const matches = (k: BlockingKeyConfig): boolean =>
      arrayEq(k.fields, this.fields) && arrayEq(k.transforms ?? [], this.transforms);
    const keys = config.blocking.keys ?? [];
    const exists = keys.some(matches);
    let nextKeys: readonly BlockingKeyConfig[];
    if (this.action === "add") {
      if (exists) return null;
      nextKeys = [
        ...keys,
        { fields: [...this.fields], transforms: [...this.transforms] },
      ];
    } else {
      if (!exists) return null;
      nextKeys = keys.filter((k) => !matches(k));
    }
    const blocking = { ...config.blocking, keys: nextKeys };
    if (!blockingValid(blocking)) return null;
    return { ...config, blocking };
  }
}

/**
 * Map one LLM-emitted edit spec (a JSON object) to a `ConfigEdit`. Returns
 * `null` for an unknown op or a malformed spec -- the closed vocabulary is the
 * only thing the LLM can drive, so junk is dropped, not run.
 */
export function editFromSpec(spec: unknown): ConfigEdit | null {
  if (typeof spec !== "object" || spec === null || Array.isArray(spec)) {
    return null;
  }
  const s = spec as Record<string, unknown>;
  const op = s["op"];
  try {
    if (op === "threshold_shift") return new ThresholdShift(asNum(s["delta"]));
    if (op === "scorer_swap") {
      return new ScorerSwap(asStr(s["matchkey"]), asStr(s["field"]), asStr(s["scorer"]));
    }
    if (op === "blocking_strategy") return new BlockingStrategyEdit(asStr(s["strategy"]));
    if (op === "weight_shift") {
      return new WeightShift(asStr(s["matchkey"]), asStr(s["field"]), asNum(s["delta"]));
    }
    if (op === "matchkey_type") {
      return new MatchkeyTypeSwap(asStr(s["matchkey"]), asStr(s["target_type"]));
    }
    if (op === "blocking_key") {
      const fields = s["fields"];
      if (!Array.isArray(fields)) return null;
      const transforms = Array.isArray(s["transforms"]) ? s["transforms"] : [];
      return new BlockingKeyEdit(
        asStr(s["action"]),
        fields.map(String),
        transforms.map(String),
      );
    }
  } catch {
    return null;
  }
  return null;
}

function asStr(v: unknown): string {
  if (v === null || v === undefined) throw new Error("missing");
  return String(v);
}

function asNum(v: unknown): number {
  const n = Number(v);
  if (!Number.isFinite(n)) throw new Error("not a number");
  return n;
}

/**
 * Parse an LLM response into validated `ConfigEdit`s. `{"action": "stop"}` /
 * `{"action": "satisfied"}` (or any non-list `edits`) yields an empty list.
 */
export function parseLlmEdits(payload: unknown): ConfigEdit[] {
  if (typeof payload !== "object" || payload === null) return [];
  const p = payload as Record<string, unknown>;
  if (p["action"] === "stop" || p["action"] === "satisfied") return [];
  const raw = p["edits"];
  if (!Array.isArray(raw)) return [];
  return raw.map(editFromSpec).filter((e): e is ConfigEdit => e !== null);
}

/**
 * Apply `edits` in sequence onto `config`, skipping any that don't apply
 * (return `null`). Returns the folded config -- equal to the input when every
 * edit was a no-op.
 */
export function foldEdits(
  config: GoldenMatchConfig,
  edits: readonly ConfigEdit[],
): GoldenMatchConfig {
  let cur = config;
  for (const edit of edits) {
    const next = edit.apply(cur);
    if (next !== null) cur = next;
  }
  return cur;
}
